<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Changelog • QBMS</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">QBMS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/bms_example.html">BMS Example</a>
    <a class="dropdown-item" href="../articles/breedbase_example.html">BreedBase Example</a>
    <a class="dropdown-item" href="../articles/ebs_example.html">EBS Example</a>
    <a class="dropdown-item" href="../articles/gigwa_example.html">GIGWA Example</a>
    <a class="dropdown-item" href="../articles/hwsd2_example.html">FAO HWSD v2 Example</a>
    <a class="dropdown-item" href="../articles/terraclimate_example.html">TerraClimate Example</a>
  </div>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/icarda-git/QBMS/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-news">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Changelog</h1>
      <small>Source: <a href="https://github.com/icarda-git/QBMS/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.5.0" id="qbms-150-2024-09-19">QBMS 1.5.0 <em>2024-09-19</em>
<a class="anchor" aria-label="anchor" href="#qbms-150-2024-09-19"></a></h2><p class="text-muted">CRAN release: 2024-09-18</p>
<ul><li>Updated the core QBMS script to replace the non-CRAN <a href="https://github.com/gaborcsardi/async" class="external-link">async</a> R package with the CRAN-standard <a href="https://cran.r-project.org/package=future" class="external-link">future</a> and <a href="https://cran.r-project.org/package=future.apply" class="external-link">future.apply</a> packages.</li>
<li>Updated the core QBMS script to substitute the deprecated <a href="https://httr.r-lib.org/" class="external-link">httr</a> R package with the recommended <a href="https://cran.r-project.org/package=httr2" class="external-link">httr2</a> package.</li>
<li>Add new <code><a href="../reference/gigwa_get_allelematrix.html">gigwa_get_allelematrix()</a></code> function that boosting the marker data retrieval speed by more than 10 times.</li>
<li>Add new <code><a href="../reference/gigwa_get_markers.html">gigwa_get_markers()</a></code> function to retrieve the marker map that describe the physical positions of the markers on the chromosomes.</li>
<li>Add new <code><a href="../reference/gigwa_get_sequences.html">gigwa_get_sequences()</a></code> function to retrieve the list of sequences in the selected GIGWA project.</li>
<li>Improve <code><a href="../reference/get_trial_obs_ontology.html">get_trial_obs_ontology()</a></code> function performance for systems supports BrAPI v2 /search/variables endpoint.</li>
<li>Improve package documentation with enhanced details.</li>
<li>Implement an internal central mechanism to handle all <a href="https://plant-breeding-api.readthedocs.io/en/latest/docs/best_practices/Search_Services.html#post-search-entity" class="external-link">BrAPI POST /search</a> endpoints, including the 202 status code or searchResultsDbId response.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="1.0.0" id="qbms-100-2024-03-07">QBMS 1.0.0 <em>2024-03-07</em>
<a class="anchor" aria-label="anchor" href="#qbms-100-2024-03-07"></a></h2><p class="text-muted">CRAN release: 2024-03-07</p>
<ul><li>Add <a href="https://ebs.excellenceinbreeding.org/" class="external-link">EBS</a> support using BrAPI v2 calls.</li>
<li>Implement a new generic internal mechanism that supports both BrAPI versions 1 and 2 calls for all core functions.</li>
<li>Support the OAuth 2.0 authentication (Authorization Code Grant flow) using the new <code><a href="../reference/login_oauth2.html">login_oauth2()</a></code> function.</li>
<li>Support external authorization option and set access token value manually using the new <code><a href="../reference/set_token.html">set_token()</a></code> function.</li>
<li>Support downloading TerraClimate netCDF data files to extract their data offline using the new <code><a href="../reference/ini_terraclimate.html">ini_terraclimate()</a></code> function.</li>
<li>Add the new <code><a href="../reference/scan_brapi_endpoints.html">scan_brapi_endpoints()</a></code> utility function to scan available BrAPI endpoints in the configured data source server.</li>
<li>Enhance the <code><a href="../reference/gigwa_get_variants.html">gigwa_get_variants()</a></code> function by adding 3 extra filtering parameters: start position, end position, and reference sequence name.</li>
<li>Fix the BreedBase login problem due to the form encoding issue by adding the new function <code><a href="../reference/login_breedbase.html">login_breedbase()</a></code> to pass the correct/expected encoding.</li>
<li>Add <a href="https://gaez.fao.org/pages/hwsd" class="external-link">HWSD v2.0</a> support to query and retrieve the FAO harmonized soil data.</li>
<li>Resolved several minor bugs reported by <a href="https://ebs.excellenceinbreeding.org/" class="external-link">EBS</a>, <a href="https://breedbase.org/" class="external-link">BreedBase</a>, and <a href="https://app.breedinginsight.net/" class="external-link">DeltaBreed</a> users for improved stability.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.1" id="qbms-091-2023-03-28">QBMS 0.9.1 <em>2023-03-28</em>
<a class="anchor" aria-label="anchor" href="#qbms-091-2023-03-28"></a></h2><p class="text-muted">CRAN release: 2023-03-28</p>
<ul><li>Hot fix the reported validation issue in the <code><a href="../reference/gigwa_set_db.html">gigwa_set_db()</a></code> function.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.9.0" id="qbms-090-2023-03-20">QBMS 0.9.0 <em>2023-03-20</em>
<a class="anchor" aria-label="anchor" href="#qbms-090-2023-03-20"></a></h2><p class="text-muted">CRAN release: 2023-03-28</p>
<ul><li>Implement caching techniques wherever applicable for faster response times to improve user experience. Thanks to <a href="https://github.com/AparicioJohan" class="external-link">Johan Steven Aparicio</a> for initiating this.</li>
<li>Improve the performance of multi-page API calls by <a href="https://github.com/gaborcsardi/async" class="external-link">optionally enabling asynchronous calls</a> to prevent blocking behavior by fetching all requested pages simultaneously.</li>
<li>Add TerraClimate support to query and retrieve climate data (including the <a href="https://www.worldclim.org/data/bioclim.html" class="external-link">19 BIOCLIM variables</a>).</li>
<li>Add a new <code><a href="../reference/get_germplasm_attributes.html">get_germplasm_attributes()</a></code> function to retrieve germplasm attributes. Thanks to <a href="https://github.com/AparicioJohan" class="external-link">Johan Steven Aparicio</a> for his contribution.</li>
<li>Add a new <code><a href="../reference/gigwa_get_metadata.html">gigwa_get_metadata()</a></code> function to retrieve metadata from GIGWA database.</li>
<li>Add new <code><a href="../reference/get_qbms_connection.html">get_qbms_connection()</a></code> and <code><a href="../reference/set_qbms_connection.html">set_qbms_connection()</a></code> functions to support connecting to multiple providers. Thanks to <a href="https://github.com/agostof" class="external-link">Francisco Agosto-Perez</a> for the suggestions.</li>
<li>The <code><a href="../reference/get_pedigree_table.html">get_pedigree_table()</a></code> function can properly detect and handle backcross cases.</li>
<li>Fix the error message when calling the <code><a href="../reference/get_pedigree_table.html">get_pedigree_table()</a></code> function if there is no case of similar genotype names. Thanks to <a href="https://github.com/AparicioJohan" class="external-link">Johan Steven Aparicio</a> for reporting this.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.8.0" id="qbms-080-2022-05-18">QBMS 0.8.0 <em>2022-05-18</em>
<a class="anchor" aria-label="anchor" href="#qbms-080-2022-05-18"></a></h2><p class="text-muted">CRAN release: 2022-05-19</p>
<ul><li>Add GIGWA support (required version 2.4.1 or higher) to query and retrieve SNPs data for GWAS and GS analysis pipelines.</li>
<li>Tiny usability enhancements (e.g., automate path parameter setting, show the server name in the login window, add progress bar to the get_program_studies function).</li>
<li>Fix mistakenly redundant location names/info returned by the get_program_studies function.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.7.0" id="qbms-070-2022-03-03">QBMS 0.7.0 <em>2022-03-03</em>
<a class="anchor" aria-label="anchor" href="#qbms-070-2022-03-03"></a></h2><p class="text-muted">CRAN release: 2022-03-02</p>
<ul><li>Add BreedBase support using BrAPI v1 calls.</li>
<li>Add functionality to get the pedigree table starting from germplasm dataset.</li>
<li>Improve set_qbms_config to generalize the way of getting the server domain from the URL.</li>
<li>Default timeout become 120 sec instead of 10.</li>
<li>Set default encoding for HTTP content to UTF-8.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.6.0" id="qbms-060-2021-10-08">QBMS 0.6.0 <em>2021-10-08</em>
<a class="anchor" aria-label="anchor" href="#qbms-060-2021-10-08"></a></h2>
<ul><li>Fix filter by year functionality in the list_trials function.</li>
<li>Fix get_germplasm_data by replaced the deprecated germplasm-search call.</li>
<li>Minimize package dependencies (rbindx replaced plyr::rbind.fill, rbindlistx replaced data.table::rbindlist, and use merge to replace dplyr::left_join).</li>
<li>Resolve compatibility issues with BrAPI changes in BMS version 19.</li>
<li>Enable to set the connection time_out in the set_qbms_config function.</li>
<li>Get entry type (test or check) in the get_germplasm_list returned data frame.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.5.0" id="qbms-050-2021-07-08">QBMS 0.5.0 <em>2021-07-08</em>
<a class="anchor" aria-label="anchor" href="#qbms-050-2021-07-08"></a></h2>
<ul><li>Fix the issue of empty list in get_germplasm_data returned results.</li>
<li>Fix retrieving error when the study has no data!</li>
<li>Enhance returned info by the get_program_studies function to include study settings and number of test/check entries.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.1" id="qbms-041-2020-10-16">QBMS 0.4.1 <em>2020-10-16</em>
<a class="anchor" aria-label="anchor" href="#qbms-041-2020-10-16"></a></h2>
<ul><li>Simplify configuration by required only the URL of the BMS login page.</li>
<li>Improve the performance of the internal get_program_trials function by passing the programDbId in the /trials GET call.</li>
<li>Add debug_qbms function to get the internal config/state object.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.4.0" id="qbms-040-2020-07-03">QBMS 0.4.0 <em>2020-07-03</em>
<a class="anchor" aria-label="anchor" href="#qbms-040-2020-07-03"></a></h2>
<ul><li>Convert it into an R package.</li>
<li>Add set_qbms_config function to setup connection configuration variables.</li>
<li>Use the double colon approach for functions from external packages.</li>
<li>Fix the deprecated API call in the get_trial_obs_ontology function.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.1" id="qbms-031-2020-06-09">QBMS 0.3.1 <em>2020-06-09</em>
<a class="anchor" aria-label="anchor" href="#qbms-031-2020-06-09"></a></h2>
<ul><li>Fix the “get_trial_data” function bug when you have more than one study in the same location.</li>
<li>Function “list_studies” returns studyName also, and function “set_study” input is studyName now.</li>
<li>Simplify the “get_germplasm_list” function output by getting rid of nested lists.</li>
<li>Deprecate the “list_all_studies” function in favor of “get_program_studies” function.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.3.0" id="qbms-030-2020-06-02">QBMS 0.3.0 <em>2020-06-02</em>
<a class="anchor" aria-label="anchor" href="#qbms-030-2020-06-02"></a></h2>
<ul><li>Call BrAPI directly (i.e. not required “CIP-RIU/brapi” from GitHub anymore).</li>
<li>Add a function to get all data of the current active trial (combined all studies).</li>
<li>Add a function to get a list of studies where given germplasm has been used.</li>
<li>Add a function to get a specific germplasm data from all program trials.</li>
<li>Handle BrAPI pagination in a proper way.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.2.0" id="qbms-020-2019-08-20">QBMS 0.2.0 <em>2019-08-20</em>
<a class="anchor" aria-label="anchor" href="#qbms-020-2019-08-20"></a></h2>
<ul><li>Adopt tidyverse style guide <a href="https://style.tidyverse.org/" class="external-link uri">https://style.tidyverse.org/</a>
</li>
<li>Add functions documentation using roxygen2 format.</li>
<li>Add basic error handling to the functions.</li>
<li>Add a function to retrieve the traits ontology of a trial.</li>
</ul></div>
    <div class="section level2">
<h2 class="pkg-version" data-toc-text="0.1.0" id="qbms-010-2019-07-24">QBMS 0.1.0 <em>2019-07-24</em>
<a class="anchor" aria-label="anchor" href="#qbms-010-2019-07-24"></a></h2>
<ul><li>Initial version.</li>
</ul></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Khaled Al-Shamaa.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

