<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="QBMS">
<title>GIGWA Example â€¢ QBMS</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GIGWA Example">
<meta property="og:description" content="QBMS">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">QBMS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.5.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/bms_example.html">BMS Example</a>
    <a class="dropdown-item" href="../articles/breedbase_example.html">BreedBase Example</a>
    <a class="dropdown-item" href="../articles/ebs_example.html">EBS Example</a>
    <a class="dropdown-item" href="../articles/gigwa_example.html">GIGWA Example</a>
    <a class="dropdown-item" href="../articles/hwsd2_example.html">FAO HWSD v2 Example</a>
    <a class="dropdown-item" href="../articles/terraclimate_example.html">TerraClimate Example</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/icarda-git/QBMS/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>GIGWA Example</h1>
                        <h4 data-toc-skip class="author">Khaled
Al-Shamaa</h4>
            
            <h4 data-toc-skip class="date">2024-09-19</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/icarda-git/QBMS/blob/HEAD/vignettes/gigwa_example.Rmd" class="external-link"><code>vignettes/gigwa_example.Rmd</code></a></small>
      <div class="d-none name"><code>gigwa_example.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="qbms">QBMS<a class="anchor" aria-label="anchor" href="#qbms"></a>
</h2>
<p>This R package assists breeders in linking data systems with their
analytic pipelines, a crucial step in digitizing breeding processes. It
supports querying and retrieving phenotypic and genotypic data from
systems like <a href="https://ebs.excellenceinbreeding.org/" class="external-link">EBS</a>, <a href="https://bmspro.io/" class="external-link">BMS</a>, <a href="https://breedbase.org" class="external-link">BreedBase</a>, and <a href="https://github.com/SouthGreenPlatform/Gigwa2" class="external-link">GIGWA</a> (using <a href="https://brapi.org/" class="external-link">BrAPI</a> calls). Extra helper functions
support environmental data sources, including <a href="https://www.climatologylab.org/terraclimate.html" class="external-link">TerraClimate</a>
and FAO <a href="https://gaez.fao.org/pages/hwsd" class="external-link">HWSDv2</a> soil
database.</p>
</div>
<div class="section level2">
<h2 id="gigwa">GIGWA<a class="anchor" aria-label="anchor" href="#gigwa"></a>
</h2>
<p><a href="https://github.com/SouthGreenPlatform/Gigwa2" class="external-link">GIGWA</a> is a
web-based tool which provides an easy and intuitive way to explore large
amounts of genotyping data by filtering the latter based not only on
variant features, including functional annotations, but also on genotype
patterns. The data storage relies on MongoDB, which offers good
scalability perspectives. GIGWA can handle multiple databases and may be
deployed in either single or multi-user mode. Finally, it provides a
wide range of popular export formats.</p>
</div>
<div class="section level2">
<h2 id="brapi">BrAPI<a class="anchor" aria-label="anchor" href="#brapi"></a>
</h2>
<p>The Breeding API (<a href="https://brapi.org/" class="external-link">BrAPI</a>) project is
an effort to enable interoperability among plant breeding databases.
BrAPI is a standardized RESTful web service API specification for
communicating plant breeding data. This community driven standard is
free to be used by anyone interested in plant breeding data
management.</p>
</div>
<div class="section level2">
<h2 id="example">
<em>Example</em><a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the QBMS library</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://icarda-git.github.io/QBMS/" class="external-link">QBMS</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The public GIGWA testing server required no authentication. If your GIGWA server </span></span>
<span><span class="co"># requires authentication, then make sure that no_auth parameter value is FALSE</span></span>
<span><span class="co"># IMPORTENT NOTE: QBMS required GIGWA version 2.4.1 or higher</span></span>
<span><span class="fu"><a href="../reference/set_qbms_config.html">set_qbms_config</a></span><span class="op">(</span><span class="st">"https://gigwa.southgreen.fr/gigwa/"</span>, </span>
<span>                time_out <span class="op">=</span> <span class="fl">300</span>, engine <span class="op">=</span> <span class="st">"gigwa"</span>, no_auth <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># If login is required, then you can use your GIGWA account (interactive mode)</span></span>
<span><span class="co"># or pass your GIGWA username and password as parameters (batch mode)</span></span>
<span><span class="co"># login_gigwa()</span></span>
<span><span class="co"># login_gigwa("gigwadmin", "nimda")</span></span>
<span></span>
<span><span class="co"># list existing databases in the current GIGWA server</span></span>
<span><span class="fu"><a href="../reference/gigwa_list_dbs.html">gigwa_list_dbs</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select a database by name</span></span>
<span><span class="fu"><a href="../reference/gigwa_set_db.html">gigwa_set_db</a></span><span class="op">(</span><span class="st">"Sorghum-JGI_v1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># list all projects in the selected database</span></span>
<span><span class="fu"><a href="../reference/gigwa_list_projects.html">gigwa_list_projects</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select a project by name</span></span>
<span><span class="fu"><a href="../reference/gigwa_set_project.html">gigwa_set_project</a></span><span class="op">(</span><span class="st">"Nelson_et_al_2011"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># list all runs in the selected project</span></span>
<span><span class="fu"><a href="../reference/gigwa_list_runs.html">gigwa_list_runs</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select a specific run by name</span></span>
<span><span class="fu"><a href="../reference/gigwa_set_run.html">gigwa_set_run</a></span><span class="op">(</span><span class="st">"run1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get a list of all samples in the selected run</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gigwa_get_samples.html">gigwa_get_samples</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show the first 6 individuals on the list of samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># query the variants (e.g., SNPs markers) in the selected run </span></span>
<span><span class="co"># that match the given criteria:</span></span>
<span><span class="co"># - max_missing: maximum missing ratio (by sample) [0-1] (default is 1 for 100%) </span></span>
<span><span class="co"># - min_maf: minimum Minor Allele Frequency (MAF) [0-1] (default is 0 for 0%) </span></span>
<span><span class="co"># - start: start position of region (zero-based, inclusive) (e.g., 19750802)</span></span>
<span><span class="co"># - end: end position of region (zero-based, exclusive) (e.g., 19850125)</span></span>
<span><span class="co"># - referenceName: reference sequence name  (e.g., '6H' in the Barley LI-AM)</span></span>
<span><span class="co"># - samples: a list of a samples subset (default is NULL will retrieve for all samples) </span></span>
<span><span class="va">marker_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gigwa_get_variants.html">gigwa_get_variants</a></span><span class="op">(</span>max_missing <span class="op">=</span> <span class="fl">0.2</span>, </span>
<span>                                    min_maf <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>                                    start <span class="op">=</span> <span class="fl">100000</span>,</span>
<span>                                    end <span class="op">=</span> <span class="fl">500000</span>,</span>
<span>                                    samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ind1"</span>, <span class="st">"ind3"</span>, <span class="st">"ind7"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Data returns in data.frame format. The first 4 columns describe attributes of the SNP </span></span>
<span><span class="co"># - rs#: variant name</span></span>
<span><span class="co"># - alleles: reference allele / alternative allele</span></span>
<span><span class="co"># - chrom: chromosome name</span></span>
<span><span class="co"># - pos: position in bp</span></span>
<span><span class="co"># while the following columns describe the SNP value for a single sample line using </span></span>
<span><span class="co"># numerical coding 0, 1, and 2 for reference, heterozygous, alternative/minor alleles.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">marker_matrix</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the metadata associated with the samples in the current active run</span></span>
<span><span class="fu"><a href="../reference/gigwa_set_db.html">gigwa_set_db</a></span><span class="op">(</span><span class="st">"DIVRICE_NB"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gigwa_set_project.html">gigwa_set_project</a></span><span class="op">(</span><span class="st">"refNB"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/gigwa_set_run.html">gigwa_set_run</a></span><span class="op">(</span><span class="st">"03052022"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get a list of all samples in the selected run</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gigwa_get_metadata.html">gigwa_get_metadata</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/View.html" class="external-link">View</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="enhanced-allele-matrix-retrieval">Enhanced Allele Matrix Retrieval<a class="anchor" aria-label="anchor" href="#enhanced-allele-matrix-retrieval"></a>
</h2>
<p>The following functions now utilize the new efficient BrAPI v2.1
/allelematrix calls, requiring version 2.6 of GIGWA or higher. This
update significantly improves the QBMS allele matrix retrieval speed,
increasing it by more than 10 times, as demonstrated by benchmark
tests.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Configure your GIGWA connection</span></span>
<span><span class="fu"><a href="../reference/set_qbms_config.html">set_qbms_config</a></span><span class="op">(</span><span class="st">"https://gigwa.southgreen.fr/gigwa/"</span>, </span>
<span>                time_out <span class="op">=</span> <span class="fl">300</span>, engine <span class="op">=</span> <span class="st">"gigwa"</span>, no_auth <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select a database by name</span></span>
<span><span class="fu"><a href="../reference/gigwa_set_db.html">gigwa_set_db</a></span><span class="op">(</span><span class="st">"Sorghum-JGI_v1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select a project by name</span></span>
<span><span class="fu"><a href="../reference/gigwa_set_project.html">gigwa_set_project</a></span><span class="op">(</span><span class="st">"Nelson_et_al_2011"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select a specific run by name</span></span>
<span><span class="fu"><a href="../reference/gigwa_set_run.html">gigwa_set_run</a></span><span class="op">(</span><span class="st">"run1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the list of all samples in the selected project</span></span>
<span><span class="va">germplasmNames</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gigwa_get_samples.html">gigwa_get_samples</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the list of all sequences in the selected project</span></span>
<span><span class="va">chroms</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gigwa_get_sequences.html">gigwa_get_sequences</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">### Get Variants Info (Geno Map) ###############################################</span></span>
<span></span>
<span><span class="op">?</span><span class="va">gigwa_get_markers</span></span>
<span><span class="va">geno_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gigwa_get_markers.html">gigwa_get_markers</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                              end <span class="op">=</span> <span class="fl">1234567</span>,</span>
<span>                              <span class="co"># chrom = c("Sb01", "Sb02"),      # chroms[1:3]</span></span>
<span>                              <span class="op">)</span></span>
<span></span>
<span><span class="co">### Get Marker Matrix ##########################################################</span></span>
<span></span>
<span><span class="op">?</span><span class="va">gigwa_get_allelematrix</span></span>
<span><span class="va">geno_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/gigwa_get_allelematrix.html">gigwa_get_allelematrix</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fl">0</span>,                  <span class="co"># default is 0</span></span>
<span>                                    end <span class="op">=</span> <span class="fl">1234567</span>,              <span class="co"># default is "" -&gt; "ref:0-"</span></span>
<span>                                    snps <span class="op">=</span> <span class="va">geno_map</span><span class="op">$</span><span class="va">`rs#`</span>,      <span class="co"># optional</span></span>
<span>                                    <span class="co"># chrom = "Sb01",           # c("Sb01", "Sb07")</span></span>
<span>                                    <span class="co"># samples = germplasmNames, # gigwa_get_samples()</span></span>
<span>                                    <span class="co"># snps_pageSize = 10000,</span></span>
<span>                                    <span class="co"># samples_pageSize = 100,</span></span>
<span>                                    <span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Khaled Al-Shamaa.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
